containers/daap-python-base/src/var/task/data_product_metadata.py

```diff
@@ -0,0 +1,117 @@
+import json
+import os
+import traceback
+
+import boto3
+import botocore
+from jsonschema import validate
+from jsonschema.exceptions import ValidationError
+
+from data_platform_logging import DataPlatformLogger
+from dataengineeringutils3.s3 import get_filepaths_from_s3_folder, read_json_from_s3
+
+s3_client = boto3.client("s3")
+
+
+# These can be moved to separate module for all paths when work starts on that
+def get_bucket_name():
+    return os.environ["BUCKET_NAME"]
+
+
+def get_data_product_metadata_path(data_product_name):
+    metadata_s3_path = os.path.join(
+        "s3://",
+        get_bucket_name(),
+        "metadata",
+        data_product_name,
+        "v1.0",
+        "metadata.json"
+    )
+    return metadata_s3_path
+
+
+def get_data_product_metadata_spec_path(version: str = None) -> str:
+    # if version is None we'll get the latest version
+    if version is None:
+        file_paths = get_filepaths_from_s3_folder(
+            os.path.join("s3://", get_bucket_name(), "data_product_metadata_spec")
+        )
+        versions = list(
+            {i for p in file_paths for i in p.split("/") if i.startswith("v")}
+        )
+        versions.sort()
+        latest_version = versions[-1]
+        spec_path = os.path.join("s3://", get_bucket_name(), "data_product_metadata_spec", latest_version, "moj_data_product_metadata_spec.json")
+    else:
+        spec_path = os.path.join("s3://", get_bucket_name(), "data_product_metadata_spec", version, "moj_data_product_metadata_spec.json")
+
+    return spec_path
+
+
+def split_bucket_and_key(path):
+    bucket, key = path.replace("s3://", "").split("/", 1)
+    return bucket, key
+
+
+class DataProductMetadata():
+    """
+    class to handle creation and updating of
+    metadata relating to a dataproduct
+    """
+    def __init__(self, data_product_name: str, logger: DataPlatformLogger):
+        self.data_product_name = data_product_name
+        self.logger = logger
+        bucket, key = split_bucket_and_key(get_data_product_metadata_path(data_product_name))
+        self.metadata_bucket = bucket
+        self.metadata_key = key
+        self._check_if_metadata_exists()
+
+
+    def _check_if_metadata_exists(self) -> bool:
+        # establish whether metadata for data product already exists
+        try:
+            # get head of object (if it exists)
+            s3_client.head_object(Bucket=self.metadata_bucket, Key=self.metadata_key)
+        except botocore.exceptions.ClientError as e:
+            if e.response['Error']['Code'] == "404":
+                self.metadata_exists = False
+            else:
+                self.logger.error(f"Uknown error - {e}")
+                raise Exception(f"Uknown error - {e}")
+        else:
+            self.logger.info("version 1 of metadata already exists for this data product")
+            self.metadata_exists = True
+
+        return self
+
+    def validate_metadata(
```

Comment:
[Minor suggestion] Since the class is called `DataProductMetadata` maybe we can drop the `_metadata` component from the method names. Then we would call it like

```python
metadata = DataProductMetadata(...)
metadata.validate(...)
metadata.write_to_s3(...)
```